File: WorldConquest/src/models/Unit.ts
// Unit Model

/**
 * Represents a military unit in an army
 */
export class Unit {
    // Unit ID
    id: string;
    
    // Unit type
    type: string;
    
    // Unit level (technology level)
    level: number;
    
    // Maximum strength (number of soldiers)
    maxStrength: number;
    
    // Current strength (number of soldiers)
    currentStrength: number;
    
    // Recruitment cost
    recruitmentCost: number;
    
    // Maintenance cost per turn
    maintenanceCost: number;
    
    // Combat effectiveness (multiplier for strength)
    combatEffectiveness: number;
    
    constructor(id: string, type: string, level: number, maxStrength: number, recruitmentCost: number, maintenanceCost: number, combatEffectiveness: number = 1.0) {
        this.id = id;
        this.type = type;
        this.level = level;
        this.maxStrength = maxStrength;
        this.currentStrength = maxStrength;
        this.recruitmentCost = recruitmentCost;
        this.maintenanceCost = maintenanceCost;
        this.combatEffectiveness = combatEffectiveness;
    }
    
    // Calculate effective combat strength
    calculateEffectiveStrength(): number {
        return Math.floor(this.currentStrength * this.combatEffectiveness);
    }
    
    // Reinforce unit to full strength
    reinforce(): number {
        const soldiersNeeded = this.maxStrength - this.currentStrength;
        const cost = Math.floor((soldiersNeeded * this.recruitmentCost) / this.maxStrength);
        this.currentStrength = this.maxStrength;
        return cost;
    }
}

// Common unit types
export class UnitTypes {
    static readonly Infantry: string = "Infantry";
    static readonly Cavalry: string = "Cavalry";
    static readonly Artillery: string = "Artillery";
    static readonly Navy: string = "Navy";
}

File: WorldConquest/Models/Unit.cs
using System.Text.Json.Serialization;

namespace WorldConquest.Models
{
    public class Unit
    {
        // Unit ID
        public string Id { get; set; }
        
        // Unit type
        public string Type { get; set; }
        
        // Unit level (technology level)
        public int Level { get; set; }
        
        // Maximum strength (number of soldiers)
        public int MaxStrength { get; set; }
        
        // Current strength (number of soldiers)
        public int CurrentStrength { get; set; }
        
        // Recruitment cost
        public int RecruitmentCost { get; set; }
        
        // Maintenance cost per turn
        public int MaintenanceCost { get; set; }
        
        // Combat effectiveness (multiplier for strength)
        public double CombatEffectiveness { get; set; }
        
        public Unit()
        {
            // Default constructor for serialization
        }
        
        public Unit(string id, string type, int level, int maxStrength, int recruitmentCost, int maintenanceCost, double combatEffectiveness = 1.0)
        {
            Id = id;
            Type = type;
            Level = level;
            MaxStrength = maxStrength;
            CurrentStrength = maxStrength;
            RecruitmentCost = recruitmentCost;
            MaintenanceCost = maintenanceCost;
            CombatEffectiveness = combatEffectiveness;
        }
        
        // Calculate effective combat strength
        public int CalculateEffectiveStrength()
        {
            return (int)(CurrentStrength * CombatEffectiveness);
        }
        
        // Reinforce unit to full strength
        public int Reinforce()
        {
            int soldiersNeeded = MaxStrength - CurrentStrength;
            int cost = (soldiersNeeded * RecruitmentCost) / MaxStrength;
            CurrentStrength = MaxStrength;
            return cost;
        }
    }
    
    // Common unit types
    public static class UnitTypes
    {
        public const string Infantry = "Infantry";
        public const string Cavalry = "Cavalry";
        public const string Artillery = "Artillery";
        public const string Navy = "Navy";
    }
}